### PR TITLE
x86_cpu_flags:Check kvm 0x40000001

### PR DESCRIPTION
**Pull request by nanliu-r**
ID: 1907773

Signed-off-by: Nana Liu <nanliu@redhat.com>

### CODE REVIEW

File: qemu/tests/x86_cpu_test_0x40000001.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+def run(test, params, env):
+    """
+    Check kvm 0x40000001 inside guest.
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    nums_cpu = params.get("smp")
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
```
Comment (vivianQizhu) [inline qemu/tests/x86_cpu_test_0x40000001.py:17]: Unnecessary.

File: qemu/tests/x86_cpu_test_0x40000001.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    session = vm.wait_for_login()
+
+    src_cpuid = os.path.join(data_dir.get_deps_dir(), "cpuid.src.tar.gz")
+    vm.copy_files_to(src_cpuid, '/home/')
+    session.cmd('tar -xzf /home/cpuid.src.tar.gz && cd /home/cpuid && make')
+    results = session.cmd_status_output('./home/cpuid/cpuid -r | grep eax=0x40000001 | uniq -c').strip()
+    if results.split()[1] != nums_cpu:
+        test.fail("some vcpu's cpuid has no eax=0x40000001.")
+    session.close()
```
Comment (vivianQizhu) [inline qemu/tests/x86_cpu_test_0x40000001.py:26]: Remember to clean files at the end.

File: qemu/tests/x86_cpu_test_0x40000001.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Check kvm 0x40000001 inside guest.
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    nums_cpu = params.get("smp")
```
Comment (vivianQizhu) [inline qemu/tests/x86_cpu_test_0x40000001.py:15]: Can not be `0` so `params['smp']` better?

File: qemu/tests/x86_cpu_test_0x40000001.py (hunk 1 above)
Comment (nanliu-r) [reply qemu/tests/x86_cpu_test_0x40000001.py]: Thanks and update.

File: qemu/tests/x86_cpu_test_0x40000001.py (hunk 2 above)
Comment (nanliu-r) [reply qemu/tests/x86_cpu_test_0x40000001.py]: added it and thanks.

File: qemu/tests/x86_cpu_test_0x40000001.py (hunk 3 above)
Comment (nanliu-r) [reply qemu/tests/x86_cpu_test_0x40000001.py]: will get smp by `vm.cpuinfo.smp`, thanks.

**Comment by nanliu-r** [comment]
(1/1) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.x86_cpu_flags.others.test_0x40000001: PASS (43.99 s)

File: qemu/tests/cfg/x86_cpu_flags.cfg [hunk 4]
```diff
@@ -109,3 +109,7 @@
                     cpu_model_flags += ",mce=off"
                     flags = ""
                     no_flags = "mce"
+                - test_0x40000001:
+                    only Linux
+                    start_vm = yes
```
Comment (vivianQizhu) [inline qemu/tests/cfg/x86_cpu_flags.cfg:114]: Maybe not necessary?

File: qemu/tests/x86_cpu_test_0x40000001.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+    nums_cpu = str(vm.cpuinfo.smp)
+
+    src_cpuid = os.path.join(data_dir.get_deps_dir(), "cpuid.src.tar.gz")
+    vm.copy_files_to(src_cpuid, '/home/')
```
Comment (vivianQizhu) [inline qemu/tests/x86_cpu_test_0x40000001.py:20]: Better define a variable for `/home` and `cpuid.src.tar.gz`, `test_dir = '/home'` `source_file = 'cpuid.src.tar.gz'`

File: qemu/tests/x86_cpu_test_0x40000001.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    try:
+        session.cmd('tar -xzf /home/cpuid.src.tar.gz -C /home/ && cd /home/cpuid-20201006 && make')
+        results = session.cmd_output('cd /home/cpuid-20201006 && ./cpuid -r | grep eax=0x40000001 | uniq -c').strip()
```
Comment (vivianQizhu) [inline qemu/tests/x86_cpu_test_0x40000001.py:23]: Also recommend a variable for the command `cmd = './cpuid -r | grep eax=0x40000001 | uniq -c'`

File: qemu/tests/cfg/x86_cpu_flags.cfg (hunk 4 above)
Comment (nanliu-r) [reply qemu/tests/cfg/x86_cpu_flags.cfg]: There has defined "start_vm = no" in Line 5.

File: qemu/tests/x86_cpu_test_0x40000001.py (hunk 5 above)
Comment (nanliu-r) [reply qemu/tests/x86_cpu_test_0x40000001.py]: Thanks and update.

File: qemu/tests/x86_cpu_test_0x40000001.py (hunk 6 above)
Comment (nanliu-r) [reply qemu/tests/x86_cpu_test_0x40000001.py]: > Also recommend a variable for the command `cmd = './cpuid -r | grep eax=0x40000001 | uniq -c'`

Thanks and update.

File: qemu/tests/x86_cpu_test_0x40000001.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+    nums_cpu = str(vm.cpuinfo.smp)
+
+    test_dir = params["test_dir"]
+    source_file = params["source_file"]
+    src_cpuid = os.path.join(data_dir.get_deps_dir(), source_file)
+    vm.copy_files_to(src_cpuid, test_dir)
+    try:
+        session.cmd('tar -xzf /home/cpuid.src.tar.gz -C /home/ && cd /home/cpuid-20201006 && make')
```
Comment (vivianQizhu) [inline qemu/tests/x86_cpu_test_0x40000001.py:24]: Here reference `test_dir` and `source_file`, I think that's why we define them.

File: qemu/tests/x86_cpu_test_0x40000001.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        check_cpuid = 'cd /home/cpuid-20201006 && ' + params["check_cpuid"]
+        results = session.cmd_output(check_cpuid).strip()
+        if results.split()[0] != nums_cpu:
+            test.fail("some vcpu's cpuid has no eax=0x40000001.")
+    finally:
+        session.cmd("rm /home/cpuid* -rf")
```
Comment (vivianQizhu) [inline qemu/tests/x86_cpu_test_0x40000001.py:30]: Also here and Line 25.

Comment (nanliu-r) [reply qemu/tests/x86_cpu_test_0x40000001.py:30]: > Also here and Line 25.
 we need to delete the source file `cpuid.src.tar.gz` and test dir `cpuid-20201006` both, would do you think the current code 
 of Line 30 is better?

Comment (vivianQizhu) [reply qemu/tests/x86_cpu_test_0x40000001.py:30]: @nanliu-r I mean `session.cmd('rm -rf %s/cpuid*' % test_dir)`

File: qemu/tests/x86_cpu_test_0x40000001.py (hunk 7 above)
Comment (nanliu-r) [reply qemu/tests/x86_cpu_test_0x40000001.py]: Thanks and update.

File: qemu/tests/x86_cpu_test_0x40000001.py (hunk 8 above)
Comment (nanliu-r) [reply qemu/tests/x86_cpu_test_0x40000001.py]: > @nanliu-r I mean `session.cmd('rm -rf %s/cpuid*' % test_dir)`

@ Thanks and update.

**Comment by nanliu-r** [comment]
Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.x86_cpu_flags.others.test_0x40000001: PASS (43.30 s)

File: qemu/tests/x86_cpu_test_0x40000001.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    nums_cpu = str(vm.cpuinfo.smp)
+
+    test_dir = params["test_dir"]
+    source_file = params["source_file"]
+    src_cpuid = os.path.join(data_dir.get_deps_dir(), source_file)
+    vm.copy_files_to(src_cpuid, test_dir)
+    guest_dir  = "%s/cpuid-20201006" % test_dir
+    try:
+        session.cmd('tar -xzf %s/cpuid.src.tar.gz -C %s && cd %s && make'
+                    % (test_dir, test_dir, guest_dir))
```
Comment (vivianQizhu) [inline qemu/tests/x86_cpu_test_0x40000001.py:26]:
```suggestion
        session.cmd('tar -xzf %s/%s -C %s && cd %s && make'
                    % (test_dir, source_file, test_dir, guest_dir))
```

Comment (nanliu-r) [reply qemu/tests/x86_cpu_test_0x40000001.py]: @vivianQizhu Thanks and update.

**Review by vivianQizhu — APPROVED** [review]

**Comment by vivianQizhu** [comment]
@nanliu-r Latest test result please.

**Comment by nanliu-r** [comment]
@vivianQizhu Test result PASS, thanks.
Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.x86_cpu_flags.others.test_0x40000001: PASS (45.49 s)

**Pull request merged**